ChopChop/Models/Recipe/Recipe/Recipe.swift

```diff
@@ -0,0 +1,204 @@
+import Foundation
+import GRDB
+
+/// Note there is no relationship between steps and ingredients after parsing stage
+class Recipe: FetchableRecord {
+    var id: Int64?
+    private(set) var name: String
+    private(set) var servings: Double
+    var recipeCategoryId: Int64?
+    private(set) var difficulty: Difficulty?
+    private(set) var steps: [RecipeStep]
+    private(set) var ingredients: [RecipeIngredient]
+
+    init(name: String, servings: Double = 1, difficulty: Difficulty? = nil, steps: [RecipeStep] = [], ingredients: [RecipeIngredient] = []) throws {
+        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedName.isEmpty else {
+            throw RecipeError.invalidName
+        }
+        self.name = trimmedName
+
+        guard servings > 0 else {
+            throw RecipeError.invalidServings
+        }
+        self.servings = servings
+        self.difficulty = difficulty
+        self.steps = steps
+        self.ingredients = ingredients
+        assert(checkRepresentation())
+    }
+
+    func updateName(_ name: String) throws {
+        assert(checkRepresentation())
+        let trimmedName = name.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedName.isEmpty else {
+            throw RecipeError.invalidName
+        }
+        self.name = trimmedName
+        assert(checkRepresentation())
+    }
+
+    func updateRecipeCategory(_ recipeCategory: Int64) {
+        assert(checkRepresentation())
+        self.recipeCategoryId = recipeCategory
+        assert(checkRepresentation())
+    }
+
+    func updateDifficulty(_ difficulty: Difficulty) {
+        assert(checkRepresentation())
+        self.difficulty = difficulty
+        assert(checkRepresentation())
+    }
+
+    // step related functions
+    func addStep(_ addedStep: RecipeStep) {
+        assert(checkRepresentation())
+        steps.append(addedStep)
+        assert(checkRepresentation())
+    }
+
+    func removeStep(_ removedStep: RecipeStep) throws {
+        assert(checkRepresentation())
+        guard (steps.contains { $0 === removedStep }) else {
+            throw RecipeError.nonExistentStep
+        }
+
+        steps.removeAll { $0 === removedStep }
+        assert(checkRepresentation())
+    }
+
+    // reorder step from larger to smaller index
+    func moveStepUp(_ movedStep: RecipeStep) throws {
```

Comment:
Is there a reason why this function and the one below work this way (shifted one space) instead of just changing the index of the `movedStep`

Reply:
I think UI wise a step will only ever index+-1 (ie step index will never be able to change from 1 to 5 immediately).
If we pass in the new index with `moveStepUp(_ movedStep: RecipeStep, newIndex: Int)`, the RecipeViewModel will have to be the one finding the index of the movedStep before passing the index to RecipeModel.

Alternatively we could have `moveStepUp(_ movedStep: RecipeStep, isSwapDown: Bool)` to combine moveUp and moveDown into 1 method